File: packages/generator-core/src/date.ts
const isoDate = /^(\d{4})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d+))?)?(Z|[+-]\d{2}:\d{2})?)?$/;
const rfcDate = /^(?:(Sun|Mon|Tue|Wed|Thu|Fri|Sat),[ \t]+)?(\d{1,2})[ \t]+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[ \t]+(\d{4})[ \t]+(\d{2}):(\d{2})(?::(\d{2}))?[ \t]+([+-]\d{4}|UT|GMT|EST|EDT|CST|CDT|MST|MDT|PST|PDT)$/i;
const months = ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec'];
const weekdays = ['sun', 'mon', 'tue', 'wed', 'thu', 'fri', 'sat'];
const zones: Readonly<Record<string, number>> = {
  UT: 0, GMT: 0, EST: -300, EDT: -240, CST: -360, CDT: -300,
  MST: -420, MDT: -360, PST: -480, PDT: -420,
};

function offsetMinutes(zone: string): number | undefined {
  if (!zone || zone === 'Z') return 0;
  const name = zone.toUpperCase();
  if (Object.prototype.hasOwnProperty.call(zones, name)) return zones[name];
  const digits = zone.slice(1).replace(':', '');
  const hours = Number(digits.slice(0, 2));
  const minutes = Number(digits.slice(2));
  if (hours > 23 || minutes > 59) return undefined;
  return (zone[0] === '-' ? -1 : 1) * (hours * 60 + minutes);
}

/** Parse the supported ISO and RFC date forms without using the host timezone. */
export function parseDateValue(value: string): Date | undefined {
  let year: number, month: number, day: number, hour: number, minute: number, second: number;
  let fraction = '', zone = '', weekday: string | undefined;
  const iso = isoDate.exec(value);
  if (iso && iso[0] === value) {
    year = Number(iso[1]); month = Number(iso[2]); day = Number(iso[3]);
    hour = Number(iso[4] ?? 0); minute = Number(iso[5] ?? 0); second = Number(iso[6] ?? 0);
    fraction = iso[7] ?? ''; zone = iso[8] ?? '';
  } else {
    const rfc = rfcDate.exec(value);
    if (!rfc || rfc[0] !== value) return undefined;
    weekday = rfc[1]?.toLowerCase();
    year = Number(rfc[4]); month = months.indexOf(rfc[3]!.toLowerCase()) + 1; day = Number(rfc[2]);
    hour = Number(rfc[5]); minute = Number(rfc[6]); second = Number(rfc[7] ?? 0); zone = rfc[8]!;
  }
  const offset = offsetMinutes(zone);
  if (offset === undefined || month < 1 || month > 12 || day < 1 || day > 31 || hour > 23 || minute > 59 || second > 59) return undefined;
  const local = new Date(0);
  local.setUTCFullYear(year, month - 1, day);
  local.setUTCHours(hour, minute, second, Number(fraction.slice(0, 3).padEnd(3, '0')));
  if (local.getUTCFullYear() !== year || local.getUTCMonth() !== month - 1 || local.getUTCDate() !== day) return undefined;
  if (weekday !== undefined && weekdays[local.getUTCDay()] !== weekday) return undefined;
  return new Date(local.getTime() - offset * 60_000);
}

/** Format UTC date parts and retain unsupported or invalid input unchanged. */
export function formatDateValue(value: string, pattern: string): string {
  const date = parseDateValue(value);
  if (!date) return value;
  const pad = (number: number) => String(number).padStart(2, '0');
  const year = date.getUTCFullYear();
  const parts: Readonly<Record<string, string>> = {
    YYYY: `${year < 0 ? '-' : ''}${String(Math.abs(year)).padStart(4, '0')}`,
    MM: pad(date.getUTCMonth() + 1), DD: pad(date.getUTCDate()),
    HH: pad(date.getUTCHours()), mm: pad(date.getUTCMinutes()), ss: pad(date.getUTCSeconds()),
  };
  return pattern.replace(/YYYY|MM|DD|HH|mm|ss/g, token => parts[token]!);
}

File: packages/generator-go/date.go
package generator

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

var isoDate = regexp.MustCompile(`^(\d{4})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d+))?)?(Z|[+-]\d{2}:\d{2})?)?$`)
var rfcDate = regexp.MustCompile(`(?i)^(?:(Mon|Tue|Wed|Thu|Fri|Sat|Sun),[ \t]+)?(\d{1,2})[ \t]+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[ \t]+(\d{4})[ \t]+(\d{2}):(\d{2})(?::(\d{2}))?[ \t]+([+-]\d{4}|UT|GMT|EST|EDT|CST|CDT|MST|MDT|PST|PDT)$`)
var dateZones = map[string]int{"UT": 0, "GMT": 0, "EST": -5 * 3600, "EDT": -4 * 3600, "CST": -6 * 3600, "CDT": -5 * 3600, "MST": -7 * 3600, "MDT": -6 * 3600, "PST": -8 * 3600, "PDT": -7 * 3600}

func dateInteger(value string) int { n, _ := strconv.Atoi(value); return n }
func dateOffset(value string) (int, bool) {
	if value == "" || value == "Z" {
		return 0, true
	}
	if offset, ok := dateZones[strings.ToUpper(value)]; ok {
		return offset, true
	}
	value = strings.ReplaceAll(value, ":", "")
	hours, minutes := dateInteger(value[1:3]), dateInteger(value[3:5])
	if hours > 23 || minutes > 59 {
		return 0, false
	}
	offset := (hours*60 + minutes) * 60
	if value[0] == '-' {
		offset = -offset
	}
	return offset, true
}

func parseDate(value string) (time.Time, bool) {
	var year, month, day, hour, minute, second int
	var zone, weekday, fraction string
	if match := isoDate.FindStringSubmatch(value); match != nil {
		year, month, day = dateInteger(match[1]), dateInteger(match[2]), dateInteger(match[3])
		hour, minute, second = dateInteger(match[4]), dateInteger(match[5]), dateInteger(match[6])
		fraction, zone = match[7], match[8]
	} else if match := rfcDate.FindStringSubmatch(value); match != nil {
		weekday, day, month, year = strings.ToLower(match[1]), dateInteger(match[2]), strings.Index("janfebmaraprmayjunjulaugsepoctnovdec", strings.ToLower(match[3]))/3+1, dateInteger(match[4])
		hour, minute, second, zone = dateInteger(match[5]), dateInteger(match[6]), dateInteger(match[7]), match[8]
	} else {
		return time.Time{}, false
	}
	if month < 1 || month > 12 || day < 1 || day > 31 || hour > 23 || minute > 59 || second > 59 {
		return time.Time{}, false
	}
	offset, ok := dateOffset(zone)
	if !ok {
		return time.Time{}, false
	}
	nanosecond := 0
	if fraction != "" {
		if len(fraction) > 9 {
			fraction = fraction[:9]
		}
		nanosecond = dateInteger(fraction + strings.Repeat("0", 9-len(fraction)))
	}
	date := time.Date(year, time.Month(month), day, hour, minute, second, nanosecond, time.FixedZone("", offset))
	if date.Year() != year || int(date.Month()) != month || date.Day() != day {
		return time.Time{}, false
	}
	if weekday != "" && strings.ToLower(date.Weekday().String()[:3]) != weekday {
		return time.Time{}, false
	}
	return date.UTC(), true
}

func dateValue(value string) string {
	if date, ok := parseDate(value); ok {
		return date.Format("2006-01-02")
	}
	return value
}
func datetimeValue(value string) string {
	if date, ok := parseDate(value); ok {
		return date.Format("2006-01-02T15:04:05")
	}
	return value
}
